### PR TITLE
fix: [CDS-79048]: support for linking servicenow configuration to change request

### PR DESCRIPTION
**Pull request by vivekkarnatiharness**
## Describe your changes
image

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/53472)

### CODE REVIEW

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/servicenow/ServiceNowTaskNgHelper.java [hunk 1]
```diff
@@ -182,8 +182,16 @@ private ServiceNowTaskNGResponse createTicketWithoutTemplate(ServiceNowTaskNGPar
...
-      String ticketUrlFromTicketId = ServiceNowUtils.prepareTicketUrlFromTicketNumber(
-          serviceNowConnectorDTO.getServiceNowUrl(), ticketNg.getNumber(), serviceNowTaskNGParameters.getTicketType());
+
+      String ticketUrlFromTicketId = "";
+      if (ticketNg.getNumber() != null) {
+        ticketUrlFromTicketId =
+            ServiceNowUtils.prepareTicketUrlFromTicketNumber(serviceNowConnectorDTO.getServiceNowUrl(),
+                ticketNg.getNumber(), serviceNowTaskNGParameters.getTicketType());
+      } else {
+        ticketUrlFromTicketId = ServiceNowUtils.prepareTicketUrlFromTicketId(serviceNowConnectorDTO.getServiceNowUrl(),
```
Comment (naman-goenka) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/task/servicenow/ServiceNowTaskNgHelper.java:192]: Can we fail elegantly if sys_id is also present

**Comment by vivekkarnatiharness** [comment]
trigger uts

**Review by abhinavmittal-harness — CHANGES_REQUESTED** [review]
> Reviewed 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/53472)* status: Some of the checks are too old. 2 file(s) to review. 1 change requests. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/servicenow/ServiceNowTaskNgHelper.java` line 193 at r1](https://reviewable.io/reviews/harness/harness-core/53472#-NfQVnZ121dIPQqsZR6L:-NfQVnZ121dIPQqsZR6M:b-d4f04p) ([raw file](https://github.com/harness/harness-core/blob/4aa4d45c7a2657479890792d401452fbad5dabde/930-delegate-tasks/src/main/java/io/harness/delegate/task/servicenow/ServiceNowTaskNgHelper.java#L193)):*
> ```Java
>       } else {
>         ticketUrlFromTicketId = ServiceNowUtils.prepareTicketUrlFromTicketId(serviceNowConnectorDTO.getServiceNowUrl(),
>             ticketNg.getFields().get("sys_id").getDisplayValue(), serviceNowTaskNGParameters.getTicketType());
> ```

Add a check for "sys_id" not being present and if that is the case add a log
Mainly in case in future we find some use case where we are not returning  the vlaue.

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/servicenow/ServiceNowTaskNgHelper.java [hunk 2]
```diff
@@ -188,9 +188,14 @@ private ServiceNowTaskNGResponse createTicketWithoutTemplate(ServiceNowTaskNGPar
         ticketUrlFromTicketId =
             ServiceNowUtils.prepareTicketUrlFromTicketNumber(serviceNowConnectorDTO.getServiceNowUrl(),
```
Comment (naman-goenka) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/task/servicenow/ServiceNowTaskNgHelper.java:189]: Two suggestions: 
1. log.info("ticketNumber created for ServiceNow: {}", ticketNg.getNumber()); (If can add a isEmpty check here, and default to sys_id)
2. Can we move this logic to ServiceNowUtils in prepareTicketUrlFromTicketNumberOrId

**Review by abhinavmittal-harness — APPROVED** [review]

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/servicenow/ServiceNowTaskNgHelper.java (hunk 2 above)
Comment (vivekkarnatiharness) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/task/servicenow/ServiceNowTaskNgHelper.java]: Done

**Comment by vivekkarnatiharness** [comment]
trigger uts

**Comment by vivekkarnatiharness** [comment]
trigger uts

**Review by naman-goenka — APPROVED** [review]

**Comment by vivekkarnatiharness** [comment]
trigger uts

**Comment by vivekkarnatiharness** [comment]
trigger messagecheck

**Review by abhinavmittal-harness — APPROVED** [review]
> Reviewed 1 of 2 files at r3, 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/53472)* status: 3 file(s) to review. 3. unresolved discussions.

**Comment by abhinavmittal-harness** [comment]
trigger review

**Comment by vivekkarnatiharness** [comment]
trigger review

**Pull request merged**